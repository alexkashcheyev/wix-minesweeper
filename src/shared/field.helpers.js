import config from '../appconfig';
import { cloneDeep } from 'lodash';


export function generateField(gameInfo) {

    // data validation
    if (!validGameInfo(gameInfo)) {
        return null;
    }

    // generate mines at random locations
    const mineCoordinates = getRandomCells(gameInfo, gameInfo.mines);

    // build an array reflecting the field
    const field = buildField(gameInfo, mineCoordinates);

    return field;
}

export function openCellsFrom(x, y, field) {

    const newField = cloneDeep(field);

    // using BFS to open cells surrounding an empty cell
    const queue = [{ x, y }];

    let maxQueueLen = 0;

    while (queue.length > 0) {
        //console.log(queue.length, 'cells in queue');
        maxQueueLen = Math.max(maxQueueLen, queue.length);

        // take first cell from the queue
        const c = queue.shift();

        // ignore already opened cells and flagged cells
        if (!newField[c.x][c.y].isOpened && !newField[c.x][c.y].isFlagged) {

            // open the current cell
            newField[c.x][c.y].isOpened = true;

            // add its neighbours to the queue if there are
            // no mines around
            if (
                newField[c.x][c.y].minesAround === 0
            ) {
                const neighbors = getNeighborsCoordinates(
                    c.x, c.y,
                    newField.length,
                    newField[c.x].length
                ).filter(
                    // This filter duplicates the if, and removing it will not 
                    // change the functionality, but it reduces the maximum reached 
                    // queue length from 4773 to 2388 for a 300x300 field with 1 mine.
                    // The "if" is still needed, because the cells still waiting in the 
                    // queue, can be already opened.
                    n => !newField[n.x][n.y].isOpened
                        && !newField[c.x][c.y].isFlagged
                );

                queue.push(...neighbors);
            }
        }

    }

    //console.log('max queue length was', maxQueueLen)

    return newField;
}

export function countClosedCells(field) {

    return field.reduce(
        (fieldSum, column) => {
            const fieldSumAddition = column.reduce(
                (colSum, cell) => {
                    const addition = (cell.isOpened ? 0 : 1);
                    return colSum + addition;
                },
                0
            )
            return fieldSum + fieldSumAddition;
        },
        0
    )

}

function validGameInfo({ width, height, mines }) {

    const hasErrors = (width < config.minWidth)
        || (width > config.maxWidth)
        || (height < config.minHeight)
        || (height > config.maxHeight)
        || (mines < config.minMines)
        || (mines > (width * height - config.minFreeCells));

    return !hasErrors;

}

function getRandomCells({ height, width}, amount) {
   
    const res = [];
    const allCells = [];
    
    // generate an aray of all possible coordinates of given field
    for (let x = 0; x < width; x++) {
        for (let y = 0; y < height; y++) {
            allCells.push({x,y});
        }
    }

    // get requested amount of coordinates by random index
    for (let i = 0; i < amount; i++) {
        const index = Math.floor(Math.random() * allCells.length);

        // removing the element at random index
        // the result of splice is an array
        const cell = allCells.splice(index, 1);
        res.push(...cell)
    }

    return res;
}

function getNeighborsCoordinates(x, y, width, height) {

    return [

        { x: -1, y: -1 }, { x: 0, y: -1 }, { x: 1, y: -1 },

        { x: -1, y:  0 }, /*   target   */ { x: 1, y:  0 },

        { x: -1, y:  1 }, { x: 0, y:  1 }, { x: 1, y:  1 }

    ].map(
        delta => ({ x: x + delta.x, y: y + delta.y })
    ).filter(
        neighbor => neighbor.x >= 0 && neighbor.x < width
            && neighbor.y >= 0 && neighbor.y < height
    );

}

function buildField({ width, height }, mineCoordinates) {
    const field = [];

    // create a 2d array filled with empty cells
    for (let x = 0; x < width; x++) {
        field[x] = [];
        for (let y = 0; y < height; y++) {
            field[x][y] = {
                isFlagged: false,
                isOpened: false,
                hasMine: false,
                minesAround: null
            }
        }
    }

    // plant the mines
    for (let m of mineCoordinates) {
        field[m.x][m.y].hasMine = true;
    }

    // calculate the numbers of mines around each cell
    for (let x = 0; x < width; x++) {
        for (let y = 0; y < height; y++) {
            
            // if the cell has mine inside, don't calculate
            if (!field[x][y].hasMine) {

                const minesAround = getNeighborsCoordinates(x, y, width, height)
                    .filter(n => field[n.x][n.y].hasMine)
                    .length;

                field[x][y].minesAround = minesAround;
            }
        }
    }

    return field;
}
