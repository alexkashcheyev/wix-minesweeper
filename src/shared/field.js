import config from '../appconfig';
import { clone } from 'lodash';

function validGameInfo({ width, height, mines }) {

    const hasErrors = (width < config.minWidth)
        || (width > config.maxWidth)
        || (height < config.minHeight)
        || (height > config.maxHeight)
        || (mines < config.minMines)
        || (mines > (width * height - config.minFreeCells));

    return !hasErrors;
}

function getRandomCells({ height, width}, amount) {
    
    const res = [];
    const allCells = [];
    
    for (let x = 0; x < width; x++) {

        for (let y = 0; y < height; y++) {

            // the biggest this array is going 
            // to be is 90000 elements
            // assuming every number is 8 bytes,
            // it will take about 700 kbytes of memory
            // but can be disposed by GC as soon as 
            // this function is finished.
            allCells.push({x,y});

        }
    }

    for (let i = 0; i < amount; i++) {
        const index = Math.floor(Math.random() * allCells.length);

        // removing the element at random index
        // the result of splice is an array
        const cell = allCells.splice(index, 1);
        res.push(...cell)
    }

    return res;
}

function getNeighborsCoordinates(x, y, width, height) {

    return [
        { x: -1, y: -1 }, { x: 0, y: -1 }, { x: 1, y: -1 },
        { x: -1, y: 0 }, { x: 1, y: 0 },
        { x: -1, y: 1 }, { x: 0, y: 1 }, { x: 1, y: 1 }
    ].map(
        delta => ({ x: x + delta.x, y: y + delta.y })
    ).filter(
        neighbor => neighbor.x >= 0 && neighbor.x < width
            && neighbor.y >= 0 && neighbor.y < height
    );

}

function buildField({ width, height }, mineCoordinates) {
    const field = [];

    for (let x = 0; x < width; x++) {
        field[x] = [];
        for (let y = 0; y < height; y++) {
            field[x][y] = {
                isFlagged: false,
                isOpened: false,
                hasMine: false,
                minesAround: null
            }
        }
    }

    for (let m of mineCoordinates) {
        field[m.x][m.y].hasMine = true;
    }

    for (let x = 0; x < width; x++) {

        for (let y = 0; y < height; y++) {

            if (!field[x][y].hasMine) {

                const minesAround = getNeighborsCoordinates(x, y, width, height)
                    .filter(n => field[n.x][n.y].hasMine)
                    .length;

                field[x][y].minesAround = minesAround;

            }

        }
    }

    console.log('ready');

    return field;
}

export function generateField(gameInfo) {

    // data validation
    if (!validGameInfo(gameInfo)) {
        return null;
    }
    // generate mines at random locations
    const mineCoordinates = getRandomCells(gameInfo, gameInfo.mines);

    // build an array reflecting the field
    // (It could take by about 2 Mb of RAM)
    const field = buildField(gameInfo, mineCoordinates);

    return field;
}


export function openCellsFrom(x, y, field) {
    const newField = clone(field);

    const queue = [{ x, y }];

    let i = 0;

    while (queue.length > 0) {
        const c = queue.shift();

        // don't open flagged cells, the user said there is a mine there!

        if (!newField[c.x][c.y].isOpened && !newField[c.x][c.y].isFlagged) {

            newField[c.x][c.y].isOpened = true;
            if (
                newField[c.x][c.y].minesAround === 0
            ) {
                const neighbors = getNeighborsCoordinates(
                    c.x, c.y,
                    newField.length,
                    newField[c.x].length
                ).filter(
                    n => !newField[n.x][n.y].isOpened
                );

                queue.push(...neighbors);
            }
        }

    }

    return newField;
}