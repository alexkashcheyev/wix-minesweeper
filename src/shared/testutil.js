import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

export function wrapComponent(unwrapped, state) {

    const store = configureStore([])(state);

    const component = (
        <Provider store={store}>
            {unwrapped}
        </Provider>
    );

    return { store, component };
}

export function countCells(field, test) {
    return field.reduce(
        (total, column) => {
            return total + column.reduce(
                (subTotal, cell) => {
                    return subTotal + (test(cell) ? 1 : 0)
                },
                0
            )
        },
        0
    );
}

export function createField(template) {


    // *        is mine
    // F        is flag
    // E        is flag on mine
    // number   is mines around

    const res = [];

    for (let y in template) {
        
        for (let x in template[y]) {

            if (typeof(res[x]) === 'undefined') {
                res[x] = [];
            }

            const token = template[y].charAt(x);

            res[x][y] = {
                isOpened: false,
                hasMine: token === '*' || token === 'E',
                isFlagged: token === 'F' || token === 'E',
                minesAround: token === '*' ? null : (parseInt(token) || 0)
            }
        }
    }

    return res;
}