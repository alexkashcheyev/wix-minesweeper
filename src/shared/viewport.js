
export function validViewportChange(viewport, gameInfo, dx, dy) {
    
    // if horizontal change requested
    if (dx !== 0) {

        // if the fields fits into the viewport, 
        // change should not be accepted
        if (viewport.width >= gameInfo.width) return false;

        // otherwise, check if we won't get over the
        // borders of the field

        const targetX = viewport.offset.x + dx;

        if (
            targetX < 0 
            || targetX + viewport.width > gameInfo.width
        ) {
            return false;
        } 
    }

    // if vertical change requested, do the same
    if (dy !== 0) {

        if (viewport.height >= gameInfo.height) return false;

        const targetY = viewport.offset.y + dy;

        if (
            targetY < 0
            || targetY + viewport.height > gameInfo.height
        ) {
            return false;
        }

    }

    return true;
}