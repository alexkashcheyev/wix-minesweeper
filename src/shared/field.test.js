import { generateField, validGameInfo, openCellsFrom, countClosedCells } from "./field";
import { countCells, createField } from '../shared/testutil';

describe('generateField', () => {

    it.each
    `
        width   | height    | mines
        ${2 }   | ${10}     | ${1}
        ${0 }   | ${10}     | ${1}
        ${10}   | ${1 }     | ${1}
        ${10}   | ${-1}     | ${1}
        ${3 }   | ${3 }     | ${9}
        ${10}   | ${10}     | ${101}
        ${301}  | ${5}      | ${50}
        ${10}   | ${500}    | ${1}
        ${5 }   | ${5 }     | ${-1}
        ${10}   | ${10}     | ${0}
    `('should return null if gameInfo is invalid', (arg) => {
        const res = generateField(arg);
        expect(res).toBeNull();
    });

    it.each
        `
        width   | height    | mines
        ${ 3 }  | ${ 3 }    | ${ 1 }
        ${ 3 }  | ${ 3 }    | ${ 8 }
        ${ 5 }  | ${ 7 }    | ${ 10}
        ${ 7 }  | ${ 5 }    | ${ 10}
        ${300}  | ${ 3 }    | ${ 1 }
        ${ 3 }  | ${300}    | ${899}
        ${300}  | ${300}    | ${ 1 }
        ${300}  | ${300}    | ${89999}
    `('should return field with expected parameters', (arg) => {
            const field = generateField(arg);
            
            const generatedMines = countCells(field, cell => cell.hasMine);

            expect(field).not.toBeNull();
            expect(field.length).toEqual(arg.width);
            expect(field[0].length).toEqual(arg.height);
            expect(generatedMines).toEqual(arg.mines);
        });

});

describe('openCellsFrom', () => {
    
    it ('should open the mine if starting from mine', () => {
        const field = createField([
            '111',
            '1*1',
            '111'
        ])

        const res = openCellsFrom(1,1, field);

        expect(res[1][1].isOpened).toBeTruthy();
        expect(res).not.toBe(field);
    })

    it ('should not change the field if started from a flagged cell', () => {
        const field = createField([
            'E10',
            '110',
            '00F'
        ]);

        // These functions must create two different fields
        // without mutating the original field,
        // still being its clones
        const mine = openCellsFrom(0,0, field);
        expect(mine).not.toBe(field);
        expect(mine).toEqual(field);

        const noMine = openCellsFrom(2,2, field);
        expect(noMine).not.toBe(field);
        expect(noMine).toEqual(field);
    })

    it ('should open only the selected cell if it has mines around', () => {
        const field = createField([
            '*10',
            '121',
            '01*'
        ]);

        [
            { x: 1, y: 0 },
            { x: 0, y: 0 },
            { x: 1, y: 1 },
            { x: 2, y: 1 },
            { x: 1, y: 2 }
        ].map(
            ({ x, y }) => openCellsFrom(x, y, field)
        ).forEach(newField => { 
            const openedCells = countCells(newField, cell => cell.isOpened)
            expect(openedCells).toEqual(1);
            expect(newField).not.toBe(field);
        });
        
    })

    it('should recursively open the cell neighbours if the cell is empty', () => {
        const field = createField([
            '*100',
            '2200',
            '*100'
        ]);

        [
            { x: 2, y: 0},
            { x: 3, y: 0 },
            { x: 2, y: 1 },
            { x: 3, y: 1 },
            { x: 2, y: 2 },
            { x: 3, y: 2 }
        ].map(
            ({x, y}) => openCellsFrom(x, y, field)
        ).forEach(newField => {
            const openedCells = countCells(newField, cell => cell.isOpened);
            expect(openedCells).toEqual(9);
            expect(newField).not.toBe(field);
        })
    })

    it('should not open cell neigbour if it is flagged', () => {
        const field = createField([
            '*F0',
            '110',
            '000'
        ])

        const res = openCellsFrom(2,2, field);
        expect(res[1][0].isOpened).toBeFalsy();
        expect(res).toMatchSnapshot();
    })

})