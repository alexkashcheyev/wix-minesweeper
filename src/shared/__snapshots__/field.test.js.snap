// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openCellsFrom should not open cell neigbour if it is flagged 1`] = `
Array [
  Array [
    Object {
      "hasMine": true,
      "isFlagged": false,
      "isOpened": false,
      "minesAround": null,
    },
    Object {
      "hasMine": false,
      "isFlagged": false,
      "isOpened": true,
      "minesAround": 1,
    },
    Object {
      "hasMine": false,
      "isFlagged": false,
      "isOpened": true,
      "minesAround": 0,
    },
  ],
  Array [
    Object {
      "hasMine": false,
      "isFlagged": true,
      "isOpened": false,
      "minesAround": 0,
    },
    Object {
      "hasMine": false,
      "isFlagged": false,
      "isOpened": true,
      "minesAround": 1,
    },
    Object {
      "hasMine": false,
      "isFlagged": false,
      "isOpened": true,
      "minesAround": 0,
    },
  ],
  Array [
    Object {
      "hasMine": false,
      "isFlagged": false,
      "isOpened": true,
      "minesAround": 0,
    },
    Object {
      "hasMine": false,
      "isFlagged": false,
      "isOpened": true,
      "minesAround": 0,
    },
    Object {
      "hasMine": false,
      "isFlagged": false,
      "isOpened": true,
      "minesAround": 0,
    },
  ],
]
`;
